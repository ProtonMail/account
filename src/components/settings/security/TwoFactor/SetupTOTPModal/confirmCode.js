import { Component } from 'preact';
import { connect } from 'unistore/full/preact';

import settingsActions from '../../../../../actions/settings';
import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';

import styles from './index.css';

export class ConfirmCode extends Component {
    constructor(props) {
        super(props);
        const {
            settings: { setupTOTP: { request: { TOTPCode } = {} } = {} }
        } = props;
        this.state = {
            code: TOTPCode
        };
    }

    componentWillReceiveProps(newProps) {
        const {
            settings: { setupTOTP: { status, error } = {} }
        } = newProps;

        if (status && status !== this.props.settings.setupTOTP.status) {
            if (status === 'success') {
                this.props.onSubmit();
            }
            if (status === 'failure') {
                this.props.onReset(error ? error.message || error : undefined);
            }
            this.setState({ loading: false });
        }
    }

    onSubmit(e) {
        e.preventDefault();
        this.setState({ loading: true });
        this.props.forbidClosure();
        this.props.enableTOTPAction(this.state.code);
    }

    render() {
        return (
            <ModalWrapper
                onSubmit={(e) => {
                    this.onSubmit(e);
                }}
                onReset={(e) => {
                    e.preventDefault();
                    this.props.onCancel();
                }}
            >
                <ModalContent>
                    <p className={styles.description}>
                        Test your new 2FA method:
                    </p>
                    <div className="form-row">
                        <label htmlFor="verifyCode">Input your code</label>
                        <div>
                            <input
                                onInput={({ target: { value: code } }) => {
                                    this.setState({ code });
                                }}
                                required={true}
                                value={this.state.code}
                                disabled={this.state.loading}
                                type="code"
                                id="verifyCode"
                                placeholder="Code"
                                minLength="6"
                                maxLength="6"
                                autoFocus={true}
                            />
                        </div>
                    </div>
                </ModalContent>
                <ModalFooter>
                    <button type="reset" value="Reset" disabled={this.state.loading}>
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={this.state.loading}>
                        Finish
                    </button>
                </ModalFooter>
            </ModalWrapper>
        );
    }
}

export default connect('settings', settingsActions)(ConfirmCode);
