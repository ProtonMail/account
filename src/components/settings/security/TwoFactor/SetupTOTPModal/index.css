.grid {
    width: 50%
}

.row {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.label {
    margin: auto 0;
    flex: 1;
}

.value {
    flex: 3
}


.result{
    flex: 2;
}

.description {
    flex: 2;
}

.code {
}
