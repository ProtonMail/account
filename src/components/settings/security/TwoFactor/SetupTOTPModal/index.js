import { h } from 'preact';

import Presentation from './presentation';
import SharedSecret from './sharedSecret';
import { steps as RecoveryCodeSteps, beforeDismiss as beforeDismissRecoveryCodeModal } from '../SaveRecoveryCodeModal';
import { steps as ScopeModal, beforeDismiss as beforeDismissScopeModal } from '../../../../auth/ScopeModal/index';
import store from '../../../../../helpers/store';
import actions from '../../../../../actions/settings';
import ConfirmCode from './confirmCode';

export const steps = [
    {
        title: 'Set Up Two Factor Authentication',
        component({ onNextStep, onPreviousStep, message }) {
            return <Presentation onSubmit={onNextStep} onCancel={onPreviousStep} message={message} />;
        }
    },
    ...ScopeModal('password'),
    {
        title: 'Set Up Two Factor Authentication',
        component({ onNextStep, onPreviousStep }) {
            return <SharedSecret onSubmit={onNextStep} onCancel={onPreviousStep} />;
        }
    },
    {
        title: 'Confirm your new method',
        component({ onNextStep, onPreviousStep, onReset, forbidClosure }) {
            return (
                <ConfirmCode
                    onSubmit={onNextStep}
                    onCancel={onPreviousStep}
                    onReset={onReset}
                    forbidClosure={forbidClosure}
                />
            );
        }
    },
    ...RecoveryCodeSteps
];

export const beforeDismiss = (success = false, reset = false) => {
    beforeDismissScopeModal();
    beforeDismissRecoveryCodeModal();
    if (success || !reset) {
        const { resetStoreAction } = actions(store);
        resetStoreAction(store.getState(), ['setupTOTP']);
    }
};
