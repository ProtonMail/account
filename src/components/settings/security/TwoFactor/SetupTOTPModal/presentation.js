import Link from '../../../../ui/Link';

import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';

/**
 * Shows information about the U2F Key usage.
 * @param {Object} props
 * @param {Function} props.onSubmit - triggers the next step.
 * @param {Function} props.onCancel - triggers the previous step.
 * @return {*}
 */
export default ({ onSubmit, onCancel, message }) => (
    <ModalWrapper
        onSubmit={(e) => {
            e.preventDefault();
            onSubmit();
        }}
        onReset={(e) => {
            e.preventDefault();
            onCancel();
        }}
    >
        {message ? (
            <ModalContent>
                <div>{message}</div>
            </ModalContent>
        ) : (
            <ModalContent>
                <p>
                    This wizard will enable Two Factor Authentication (2FA) on your ProtonMail account. 2FA will make
                    your ProtonMail account more secure so we recommend enabling it.{' '}
                </p>
                <p>
                    <strong>
                        If you have never used 2FA before, we strongly recommend you reading our 2FA Guide first.
                    </strong>
                </p>
                <div>
                    <Link href="https://protonmail.com/support/knowledge-base/two-factor-authentication/">
                        2FA GUIDE
                    </Link>
                </div>
            </ModalContent>
        )}
        <ModalFooter>
            <button type="reset" value="Reset">
                Cancel
            </button>
            <button type="submit" value="Submit">
                Next
            </button>
        </ModalFooter>
    </ModalWrapper>
);
