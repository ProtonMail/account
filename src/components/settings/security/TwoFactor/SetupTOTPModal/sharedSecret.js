import { Component } from 'preact';
import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';
import TextButton from '../../../../ui/TextButton';
import QRCode from 'qrcode.react';

import { connect } from 'unistore/full/preact';
import settingsActions from '../../../../../actions/settings';

import styles from './index.css';

export class SharedSecret extends Component {

    componentDidMount() {
        this.props.createSharedSecretAction();
    }

    constructor(props) {
        super(props);
        this.state = {
            showingQRCode: true
        };
    }

    renderSwitchModeButton() {
        return (<p>
            <TextButton
                onClick={() => this.setState({ showingQRCode: !this.state.showingQRCode })}>
                {this.state.showingQRCode ? 'Enter key manually instead' : 'Scan QR code'}
            </TextButton>
        </p>);
    }

    renderQRCode() {
        const { setupTOTP: { request: { qrURI } = {} } = {} } = this.props.settings;

        return (<ModalContent>
            <p>{this.props.message || 'Scan this QR code with your two factor authentication device to set up your account. '}</p>
            {this.renderSwitchModeButton()}
            {qrURI
                ? <QRCode value={qrURI} renderAs={'svg'} fgColor={'#505061'} size={256}/>
                : <p>Loading...</p>
            }
        </ModalContent>);

    }

    renderRawInformation() {
        const { setupTOTP: { request: { interval, digits, secret } = {} } = {} } = this.props.settings;

        return <ModalContent>
            <p>{this.props.message || 'Manually enter this information into your two factor authentication device to set up your account. '}</p>
            {this.renderSwitchModeButton()}

            {secret ?
                (<div class={styles.grid}>
                    <div class={styles.row}>
                        <label class={styles.label}>KEY</label>
                        <span class={styles.value}><pre>{secret}</pre></span>
                    </div>
                    <div class={styles.row}>
                        <label class={styles.label}>INTERVAL</label>
                        <span class={styles.value}><pre>{interval} seconds</pre></span>
                    </div>
                    <div class={styles.row}>
                        <label class={styles.label}>LENGTH</label>
                        <span class={styles.value}><pre>{digits} digits</pre></span>
                    </div>
                </div>)
                : <p>Loading...</p>
            }
        </ModalContent>;
    }

    render() {
        return (
            <ModalWrapper
                onSubmit={(e) => {
                    e.preventDefault();
                    this.props.onSubmit();
                }}
                onReset={(e) => {
                    e.preventDefault();
                    this.props.onCancel();
                }}
            >
                {this.state.showingQRCode ? this.renderQRCode() : this.renderRawInformation()}
                <ModalFooter>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" value="Submit">
                        Next
                    </button>
                </ModalFooter>
            </ModalWrapper>
        );
    }
}

export default connect('settings', settingsActions)(SharedSecret);
