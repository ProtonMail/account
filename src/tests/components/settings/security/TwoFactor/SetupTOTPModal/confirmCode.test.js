import render from 'preact-render-to-string';
import { deep } from 'preact-render-spy';

import { ConfirmCode } from '../../../../../../components/settings/security/TwoFactor/SetupTOTPModal/confirmCode';

const updateTreeStatus = (props) => {
    const code = 425697;

    const onSubmit = jest.fn();
    const onReset = jest.fn();
    const forbidClosure = jest.fn();
    const enableTOTPAction = jest.fn();
    const event = { preventDefault: () => undefined };

    const tree = deep(
        <ConfirmCode
            settings={{
                setupTOTP: {
                    request: {
                        TOTPCode: code
                    },
                    status: 'init'
                }
            }}
            onSubmit={onSubmit}
            onReset={onReset}
            forbidClosure={forbidClosure}
            enableTOTPAction={enableTOTPAction}
        />
    );
    tree.find('form').simulate('submit', event); // using the loading state

    expect(tree).toMatchSnapshot();

    tree.render(
        <ConfirmCode
            settings={{
                setupTOTP: {
                    request: {
                        TOTPCode: code
                    },
                    ...props
                }
            }}
            onSubmit={onSubmit}
            onReset={onReset}
        />
    );

    return {
        tree,
        onSubmit,
        onReset
    };
};

describe('SetupTOTPModal confirmCode step', () => {
    test('initial display', () => {
        expect(render(<ConfirmCode settings={{}} />)).toMatchSnapshot();
    });

    test('change code', () => {
        const tree = deep(<ConfirmCode settings={{}} />);
        tree.find('input').simulate('input', { target: { value: 424242 } });
        expect(tree).toMatchSnapshot();
    });

    test('submit and cancel', () => {
        const code = 425697;

        const onSubmit = jest.fn();
        const onCancel = jest.fn();
        const forbidClosure = jest.fn();
        const enableTOTPAction = jest.fn();

        const tree = deep(
            <ConfirmCode
                settings={{
                    setupTOTP: {
                        request: {
                            TOTPCode: code
                        }
                    }
                }}
                onCancel={onCancel}
                onSubmit={onSubmit}
                forbidClosure={forbidClosure}
                enableTOTPAction={enableTOTPAction}
            />
        );
        const event = { preventDefault: () => undefined };

        tree.find('form').simulate('reset', event);
        expect(onCancel).toBeCalled();
        expect(onSubmit).not.toBeCalled();
        expect(forbidClosure).not.toBeCalled();
        expect(enableTOTPAction).not.toBeCalled();

        tree.find('form').simulate('submit', event);
        expect(onCancel).toHaveBeenCalledTimes(1); // last event
        expect(onSubmit).not.toBeCalled();
        expect(forbidClosure).toBeCalled();
        expect(enableTOTPAction).toBeCalled();
        expect(enableTOTPAction).toHaveBeenCalledWith(code);
    });

    test('receiving status success', () => {
        const { tree, onSubmit } = updateTreeStatus({ status: 'success' });

        expect(tree).toMatchSnapshot();
        expect(onSubmit).toHaveBeenCalledTimes(1);

        // not changing the state
        tree.render(
            <ConfirmCode
                settings={{
                    setupTOTP: {
                        request: {
                            TOTPCode: '132456'
                        },
                        status: 'success'
                    }
                }}
                onSubmit={onSubmit}
            />
        );
        expect(onSubmit).toHaveBeenCalledTimes(1);
    });

    test('receiving status failure and no error', () => {
        const { tree, onReset } = updateTreeStatus({ status: 'failure' });

        expect(tree).toMatchSnapshot();
        expect(onReset).toHaveBeenCalledTimes(1);
        expect(onReset).toHaveBeenCalledWith(undefined);

        // not changing the state
        tree.render(
            <ConfirmCode
                settings={{
                    setupTOTP: {
                        request: {
                            TOTPCode: '132456'
                        },
                        status: 'failure'
                    }
                }}
                onReset={onReset}
            />
        );
        expect(onReset).toHaveBeenCalledTimes(1);
    });
    test('receiving status failure and text error', () => {
        const errorMessage = 'This is the error message';
        const { tree, onReset } = updateTreeStatus({ status: 'failure', error: errorMessage });

        expect(tree).toMatchSnapshot();
        expect(onReset).toHaveBeenCalledTimes(1);
        expect(onReset).toHaveBeenCalledWith(errorMessage);
    });

    test('receiving status failure native error', () => {
        const errorMessage = 'This is the error message';
        const { tree, onReset } = updateTreeStatus({ status: 'failure', error: new Error(errorMessage) });

        expect(tree).toMatchSnapshot();
        expect(onReset).toHaveBeenCalledTimes(1);
        expect(onReset).toHaveBeenCalledWith(errorMessage);
    });

    test('receiving props without status', () => {
        const code = 425697;

        const onSubmit = jest.fn();
        const onReset = jest.fn();
        const forbidClosure = jest.fn();
        const enableTOTPAction = jest.fn();
        const event = { preventDefault: () => undefined };

        const tree = deep(
            <ConfirmCode
                settings={{
                    setupTOTP: {
                        request: {
                            TOTPCode: code
                        },
                        status: 'init'
                    }
                }}
                onSubmit={onSubmit}
                onReset={onReset}
                forbidClosure={forbidClosure}
                enableTOTPAction={enableTOTPAction}
            />
        );
        tree.find('form').simulate('submit', event); // using the loading state

        expect(tree).toMatchSnapshot();

        tree.render(<ConfirmCode settings={{}} onSubmit={onSubmit} onReset={onReset} />);

        expect(onReset).not.toBeCalled();
        expect(onSubmit).not.toBeCalled();
    });
});
