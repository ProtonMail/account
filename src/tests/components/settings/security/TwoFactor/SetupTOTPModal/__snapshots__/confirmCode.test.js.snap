// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupTOTPModal confirmCode step change code 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={424242}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step initial display 1`] = `
<form class="wrapper">
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input required type="code" id="verifyCode" placeholder="Code" minLength="6"
        maxLength="6" autoFocus />
      </div>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Finish</button>
  </div>
</form>
`;

exports[`SetupTOTPModal confirmCode step receiving props without status 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={true}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={true}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={true}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status failure and no error 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={true}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={true}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={true}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status failure and no error 2`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={false}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={false}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status failure and text error 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={true}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={true}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={true}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status failure and text error 2`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={false}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={false}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status failure native error 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={true}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={true}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={true}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status failure native error 2`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={false}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={false}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status success 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={true}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={true}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={true}
    >
      Finish
    </button>
  </div>
</form>

`;

exports[`SetupTOTPModal confirmCode step receiving status success 2`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <p class="description">Test your new 2FA method:</p>
    <div class="form-row">
      <label htmlFor="verifyCode">Input your code</label>
      <div>
        <input
          onInput={[Function onInput]}
          required={true}
          value={425697}
          disabled={false}
          type="code"
          id="verifyCode"
          placeholder="Code"
          minLength="6"
          maxLength="6"
          autoFocus={true}
         />
      </div>
    </div>
  </div>
  <div class="footer">
    <button
      type="reset"
      value="Reset"
      disabled={false}
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Finish
    </button>
  </div>
</form>

`;
